docs/design/2019-11-05-index-advisor.md

```diff
@@ -0,0 +1,160 @@
+
+----------------------------------------------
+### A quick exmaple for single-column index recommendation
+
+**Workload**:
+
+```
+- Query1: SELECT COUNT(*) FROM t1 where a = 1 and b = 5
+- Query2: SELECT a FROM t1 ORDER BY c limit 10 
+- Query3: DELETE FROM t1 WHERE a = 1 AND d = 1
+```
+where index on column *c* has already existed. 
+
+- **Step1**: recommend indexes for single query. For each query, compile it and get the `original_cost` output by SQL optimizer. Then add all possible hypothetic indexes *a, b, d* to the schema, compile the same query again. After the compiling, `LSIs` can be extracted from output execution plan as well as the `virtual_cost`. Denote `reduced_cost` as *original_cost - virtual_cost*. Assume this step outputs:
+
+```
+Query 1--LSIs: {a, b}, reduced_cost: 4
+Query 2--LSIs: {c},    reduced_cost: 4
+Query 3--LSIs: {a, d}, reduced_cost: 3
+```
+- **Step2**: get the score list. Simply add up `LSIs`' `reduced_cost` of each query respectively and then sort the list descending by score. Assume this step outputs:
+
+```
+a: 4+3=7
+b: 4
+c: 4
+d: 3
+```
+
+- **Step3**: evaluate recommend index set. Assume the number of allowed indexes recommended by index advisor is *2* and the swap size is *1* 
+- The initial `recommend index set` is `{a, b}`, the `remaining index set` is `{c,d}` 
+- Add recommend index set `{a, b}` to the schema and at the same time mute existed index `c`
+- Compile all queries in the workload and add up each query's cost. Assume the sum is *20* and then assign it variable `min_cost`
+- Randomly exchange *1* element between the recommend set and remaining set. Assume *a* and *c* is chosen respectively, now the recommend set is `{c, b}` and the remaining set is `{a, d}`. 
+- Apply recommend index set `{c, b}` to the workload and get a sum cost, say *18*. 
+- As the cost decreased from *20* to *18*, update the recommend set from `{a, b}` to `{c, b}` and update `min_cost` to *18*. Then try swapping 1 element between `{c, b}` and `{a, d}` and execute the re-evaluate process. 
+
+Note that if the sum cost with `{c, b}` is *22* ( greater than 20 ), the next *swap* process should happen between `{a, b}` and `{c, d}` instead of between `{c, b}` and `{a, d}`. Loop the *Swap and Re-evaluate* process until timeout.
+
+## Rational 
+
+- Index advisor for PostgreSQL: [official page](https://www.enterprisedb.com/edb-docs/d/edb-postgres-advanced-server/user-guides/user-guide/9.6/EDB_Postgres_Advanced_Server_Guide.1.55.html), [discussions on index advise tool](https://www.postgresql.org/message-id/CA%2BwPC0OG5Ww2TT1bVbm8B3jhj0yY3dxhmMeOt-EP1DttoXNJuw%40mail.gmail.com)
+
+- [IBM-DB2 index advisor](https://www.ibm.com/support/knowledgecenter/en/SSEPGG_10.1.0/com.ibm.db2.luw.admin.cmd.doc/doc/r0002452.html)
+
+## Compatibility
```

Comment:
Compatibility is some thing like 'Does this proposal make TiDB not compatible with the old versions?', now it looks like implementations.
You can just leave it as no compatibility issues.

Reply:
OK